File: Java/Solution_125.java
class Solution_125 {
    public boolean isPalindrome(String s) {
        StringBuilder normalized = new StringBuilder();
        for (char c : s.toCharArray()) {
            if(Character.isLetterOrDigit(c)){
                normalized.append(Character.toLowerCase(c));
            }
        }

        String normalizedStr = normalized.toString();

        int left = 0;
        int right = normalized.length() - 1;

        while(left < right){
            if (normalizedStr.charAt(left) != normalizedStr.charAt(right)){
                return false;
            }
            left++;
            right--;
        }

        return true;
    }
    public static void main(String[] args) {
        Solution_125 sol = new Solution_125();

        String s1 = "A man, a plan, a canal: Panama";
        System.out.println(sol.isPalindrome(s1)); // Output: true

        String s2 = "race a car";
        System.out.println(sol.isPalindrome(s2)); // Output: false

        String s3 = " ";
        System.out.println(sol.isPalindrome(s3)); // Output: true
    }
}

File: Python/Solution_125.py
class Solution_125(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        normalized = []
        
        for char in s:
            if char.isalnum():
                normalized.append(char.lower())
                
        left, right = 0, len(normalized) - 1
        
        while left < right:
            if normalized[left] != normalized[right]:
                return False
            left += 1
            right -= 1
            
        return True
    
if __name__ == "__main__":
    solution = Solution_125()
    
    s1 = "A man, a plan, a canal: Panama"
    s2 = "race a car"
    s3 = " "

    print(solution.isPalindrome(s1)) # Expected Output: true
    print(solution.isPalindrome(s2)) # Expected Output: false
    print(solution.isPalindrome(s3)) # Expected Output: true